### ISSUE TITLE
Rule S4462: Calls to 'async' methods should not be blocking

### ISSUE THREAD

**Issue by alexangas** [SonarSource/sonar-csharp#1159]
### Description

Code base has uses of `.Result` when it should be using `await` to [prevent deadlocks](https://msdn.microsoft.com/en-us/magazine/jj991977.aspx).

I can't find a rule anywhere that can pick this up. Is it something that fits within the scope of this project, or should it be a custom rule?

### Repro steps

Please provide the steps required to reproduce the problem

1. Analyse code with async calls using `.Result`

2. Review analysis

### Expected behavior

Bug or code smell found where it is recommended to use `await` instead of `.Result`.

### Actual behavior

No issue found.

### Known workarounds

Writing a custom rule may be a workaround.

### Related information

* SonarC# Version: unknown (I don't have access)
* Visual Studio Version: 2017
* If running through the Scanner for MSBuild, its version: 3.0.2

**Comment by Evangelink** [comment]
Hi @alexangas,

Good suggestion! I used the article as a way to extend the rule and we came up with this rule: https://jira.sonarsource.com/browse/RSPEC-4462.

Thank you!

**Comment by alexangas** [comment]
Thank you - that is super responsive! Love your work.

**Comment by Corniel** [comment]
Actual test might look a bit like this:

``` C#
using System.Threading.Tasks;

namespace SonarAnalyzer.UnitTest.TestCases
{
    public class AwaitAsyncFunctionCall
    {
        public boo Run()
        {
            SomeAction(); // Not compliant
            return SomeFunctionAsync().Result; // Not compliant
        }

        public bool RunSync()
        {
            Task.Run(() => SomeAction()).Wait(); // Compliant
            return Task.Run(() => SomeFunctionAsync()).Result; // compliant
        }

        public async Task<bool> RunAsync()
        {
            await SomeAction().ConfigureAwait(false); // Compliant
            return await SomeFunctionAsync().ConfigureAwait(false); // Compliant
        }

        public Task<bool> SomeFunctionAsync() => Task.FromResult(true);

        public Task SomeAction() => Task.CompletedTask;
    }
}
```

Although what to do when you can't propagate the async to the calling method is open for debate.

**Comment by Evangelink** [comment]
Hi @Corniel,

Are you sure that calling a blocking method when using the pattern `Task.Run(() => FooAsync()).BlockingMethod` is actually working properly without deadlocks?

**Comment by Corniel** [comment]
@Evangelink : I linked a blog (in the duplicated topic #1197) to the code below.

That is the way how Microsoft does it. Unfortunately, that code (Microsoft's version) is internal only. It is not the responsibility (nor possible) to SonarLint to provide such a async helper.

That being said, a lot of people suggest that Task.Run(()=> SomeFunctionAsync()).Result should be sufficient. I have my doubts, but no proof yet that I'm right. I think it proves the point that you want guidance when dealing with async code that (for some unfortunate reason) can not be called by propagating the async signature to the caller. 

``` C#
using System;
using System.Threading;
using System.Threading.Tasks;

/// <summary>Helps with running async code synchronously.</summary>
/// <remarks>
/// For this one we have a helper class borrowed from Microsoft. It appears
/// is various namespaces, but always as an internal, so you can't just use
/// it directly from the framework.
/// 
/// See also: https://cpratt.co/async-tips-tricks/
/// </remarks>
public static class AsyncHelper
{
	private static readonly TaskFactory factory = new TaskFactory(
		CancellationToken.None,
		TaskCreationOptions.None,
		TaskContinuationOptions.None,
		TaskScheduler.Default);

	/// <summary>Runs the async <see cref="Task{TResult}"/> synchronously.</summary>
	/// <typeparam name="TResult">
	/// The result type of the <see cref="Task{TResult}"/>.
	/// </typeparam>
	/// <param name="func">
	/// THe function to run.
	/// </param>
	/// <returns>
	/// The result of the executed task.
	/// </returns>
	public static TResult RunSync<TResult>(Func<Task<TResult>> func)
	{
		Guard.NotNull(func, nameof(func));

		return factory
			.StartNew(func)
			.Unwrap()
			.GetAwaiter()
			.GetResult();
	}

	/// <summary>Runs the async <see cref="Task"/> synchronously.</summary>
	/// <param name="func">
	/// THe function to run.
	/// </param>
	public static void RunSync(Func<Task> func)
	{
		Guard.NotNull(func, nameof(func));

		factory
			.StartNew(func)
			.Unwrap()
			.GetAwaiter()
			.GetResult();
	}
}
```

**Comment by alexangas** [comment]
I must admit, the depths of this topic is beyond me. However Stephen Cleary has a good reputation (he wrote the linked MSDN Magazine article) and has a [lot of writings](https://blog.stephencleary.com/) about it. There is also his https://github.com/StephenCleary/AsyncEx library. It would be interesting to run the implemented SonarC# rule associated with this ticket against that library to verify correctness. 😄
